### PR TITLE
add cross_kernel.cu

### PR DESCRIPTION
**Pull request by zhangboSJTU**
### PR types

Performance optimization
### PR changes

OPs
### Describe

- add cross_kernel.cu & cross_grad_kernel.cu for cross calculation on GPU
- merge the cross_kernel.cc and cross_kernel_impl.h
- merge the cross_grad_kernel.cc and cross_grad_kernel_impl.h

### CODE REVIEW

File: paddle/phi/kernels/gpu/cross_kernel.cu [hunk 1]
```diff
@@ -13,10 +13,110 @@
...
+  __global__ void Cross(const T* x, const T* y, T* out, int outer_loops, int slice_size) {
+    unsigned int i = blockIdx.x * blockDim.x + threadIdx.x;
+    unsigned int k = blockIdx.y * blockDim.y + threadIdx.y;
+    
+    if(i < outer_loops && k < slice_size){
+      auto pos0 = (3 * i + 0) * slice_size;
+      auto pos1 = (3 * i + 1) * slice_size;
+      auto pos2 = (3 * i + 2) * slice_size;
+
+      out[pos0 + k] = x[pos1 + k] * y[pos2 + k] - x[pos2 + k] * y[pos1 + k];
```
Comment (ZzSean) [inline paddle/phi/kernels/gpu/cross_kernel.cu]: 既然都要`+k`可以在计算pos时都加上吧

File: paddle/phi/kernels/gpu/cross_kernel.cu [hunk 2]
```diff
@@ -13,10 +13,110 @@
...
+    auto slice_size = 1;
+    for (auto i = dim + 1; i < input_x_dims.size(); i++) {
+      slice_size *= input_x_dims[i];
+    }
+
+    const auto* input_x_data = input_x.data<T>();
+    const auto* input_y_data = input_y.data<T>();
+    auto* out_data = dev_ctx.template Alloc<T>(out);
+
+    dim3 threads(16, 16);
```
Comment (ZzSean) [inline paddle/phi/kernels/gpu/cross_kernel.cu]: 这个线程配置效率可能不高

File: paddle/phi/kernels/gpu/cross_kernel.cu [hunk 3]
```diff
@@ -13,10 +13,110 @@
...
 #include "paddle/phi/backends/gpu/gpu_context.h"
 #include "paddle/phi/core/kernel_registry.h"
 
+namespace phi {
+  template <typename T>
+  __global__ void Cross(const T* x, const T* y, T* out, int outer_loops, int slice_size) {
+    unsigned int i = blockIdx.x * blockDim.x + threadIdx.x;
+    unsigned int k = blockIdx.y * blockDim.y + threadIdx.y;
+    
+    if(i < outer_loops && k < slice_size){
```
Comment (ZzSean) [inline paddle/phi/kernels/gpu/cross_kernel.cu]: 这里需要加循环，如果shape过大，grid超出最大值，结果会出错

File: paddle/phi/kernels/gpu/cross_kernel.cu [hunk 4]
```diff
@@ -13,10 +13,110 @@
...
+    PADDLE_ENFORCE_EQ(
+        dims_match,
+        true,
+        phi::errors::InvalidArgument("The 'shape' of Input(X) should be equal to "
+                                     "the 'shape' of Input(Y). But received "
+                                     "Input(X).dimensions = [%s], "
+                                     "Input(Y).dimensions = [%s]",
+                                     input_x_dims,
+                                     input_x_dims));
+
```
Comment (JamesLim-sy) [inline paddle/phi/kernels/gpu/cross_kernel.cu]: L51~L63这部分的判断可以直接用下列的代替，`dims()` 的`==`应该是做了重载的：
```
PADDLE_ENFORCE_EQ(
        input_x.dims(),
        input_y.dims(),
        phi::errors::InvalidArgument("The 'shape' of Input(X) should be equal to the 'shape' of Input(Y). But received the dimension of X is [%s],  the dimension of Y is [%s]",  input_x.dims(),  input_y.dims()));
```
这块的条件判断代码可以放入InferShape 中：
```
class CrossGradOp : public framework::OperatorWithKernel {
 public:
  using framework::OperatorWithKernel::OperatorWithKernel;

  void InferShape(framework::InferShapeContext* ctx) const override {
    PADDLE_ENFORCE_EQ(
        ctx->HasInput("X"), true,
        platform::errors::InvalidArgument("Input(X) should be not null."));
    PADDLE_ENFORCE_EQ(
        ctx->HasInput("Y"), true,
        platform::errors::InvalidArgument("Input(Y) should be not null."));
    PADDLE_ENFORCE_EQ(ctx->HasInput(framework::GradVarName("Out")), true,
                      platform::errors::InvalidArgument(
                          "Input(Out@GRAD) should be not null."));
    PADDLE_ENFORCE_EQ(ctx->HasOutput(framework::GradVarName("X")), true,
                      platform::errors::InvalidArgument(
                          "Output(X@GRAD) should be not null."));
    PADDLE_ENFORCE_EQ(ctx->HasOutput(framework::GradVarName("Y")), true,
                      platform::errors::InvalidArgument(
                          "Output(Y@GRAD) should be not null."));

    ctx->SetOutputDim(framework::GradVarName("X"), ctx->GetInputDim("X"));
    ctx->SetOutputDim(framework::GradVarName("Y"), ctx->GetInputDim("Y"));
  }

 protected:
  framework::OpKernelType GetExpectedKernelType(
      const framework::ExecutionContext& ctx) const override {
    return framework::OpKernelType(OperatorWithKernel::IndicateVarDataType(
                                       ctx, framework::GradVarName("Out")),
                                   ctx.device_context());
  }
};
```

File: paddle/phi/kernels/cpu/cross_kernel.cc [hunk 5]
```diff
@@ -13,10 +13,109 @@
 // limitations under the License.
 
 #include "paddle/phi/kernels/cross_kernel.h"
-#include "paddle/phi/kernels/impl/cross_kernel_impl.h"
+#include "paddle/fluid/framework/tensor_util.h"
+#include "paddle/phi/core/dense_tensor.h"
+#include "paddle/phi/kernels/funcs/common_shape.h"
 
 #include "paddle/phi/backends/cpu/cpu_context.h"
 #include "paddle/phi/core/kernel_registry.h"
```
Comment (JamesLim-sy) [inline paddle/phi/kernels/cpu/cross_kernel.cc:20]: 头文件可以看看有没有可以缩略的空间

File: paddle/phi/kernels/gpu/cross_kernel.cu (hunk 1 above)
Comment (JamesLim-sy) [reply paddle/phi/kernels/gpu/cross_kernel.cu]: 这块计算的核心瓶颈是memry-bound，可以尝试些优化这种情况的方法。

Comment (zhangboSJTU) [reply paddle/phi/kernels/gpu/cross_kernel.cu]: kernel中重新修改线程的分配

File: paddle/phi/kernels/gpu/cross_kernel.cu (hunk 2 above)
Comment (zhangboSJTU) [reply paddle/phi/kernels/gpu/cross_kernel.cu]: > 这个线程配置效率可能不高

修改为paddle统一的线程配置，并将线程改为一维度线程

File: paddle/phi/kernels/gpu/cross_kernel.cu (hunk 3 above)
Comment (zhangboSJTU) [reply paddle/phi/kernels/gpu/cross_kernel.cu]: > 这里需要加循环，如果shape过大，grid超出最大值，结果会出错

同上，修改线程配置

File: paddle/phi/kernels/gpu/cross_kernel.cu (hunk 4 above)
Comment (zhangboSJTU) [reply paddle/phi/kernels/gpu/cross_kernel.cu]: > 

ok

File: paddle/phi/kernels/cpu/cross_kernel.cc (hunk 5 above)
Comment (zhangboSJTU) [reply paddle/phi/kernels/cpu/cross_kernel.cc:20]: 好的，这个检查过了，再看一下。

**Comment by zhangboSJTU** [comment]
> [CLA assistant check](https://cla-assistant.io/PaddlePaddle/Paddle?pullRequest=43182) Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/PaddlePaddle/Paddle?pullRequest=43182) before we can accept your contribution.**1** out of **2** committers have signed the CLA.✅ zhangbopd❌ zhangbo63
> **zhangbo63** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).
> You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/PaddlePaddle/Paddle?pullRequest=43182) it.

之前设置用户名出问题，用户名不存在，无法签署

**Pull request closed without merging**